2-repositories/scala-repo-server/src/main/scala/com.snowplowanalytics.iglu.server/service/SchemaService.scala

```diff
@@ -45,96 +49,100 @@ import javax.ws.rs.Path
+      case None => Left(AuthenticationFailedRejection(CredentialsRejected, HttpChallenges.basic("")))
+    }
+    onSuccess(credentialsRequest).flatMap {
+      case Right(user) => provide(user)
+      case Left(rejection) => reject(rejection)
+    }
+  }
 
   /**
    * Directive to validate the schema provided (either by query param or form
    * data) is self-describing.
    */
   def validateSchema(format: String): Directive1[String] =
-    anyParam('schema) | entity(as[String]) flatMap { schema =>
+    formField('schema) | parameter('schema) | entity(as[String]) flatMap { schema =>
       onSuccess((schemaActor ? ValidateSchema(schema, format))
-          .mapTo[(StatusCode, String)]) flatMap { ext =>
-            ext match {
+          .mapTo[(StatusCode, String)]) tflatMap  {
               case (OK, j) => provide(j)
-              case res => complete(res)
+              case rej => complete(rej)
             }
       }
-    }
 
   /**
    * Schema service's route
    */
-  lazy val routes =
+  lazy val routes: Route =
     rejectEmptyResponse {
-      respondWithMediaType(`application/json`) {
-        auth { authPair =>
-          post {
-            addRoute(authPair._1, authPair._2)
-          } ~
-          put {
-            updateRoute(authPair._1, authPair._2)
-          } ~
-          delete {
-            deleteRoute(authPair._1, authPair._2)
-          } ~
-          get {
-            path("public") {
-              publicSchemasRoute(authPair._1, authPair._2)
+      (get | post | put | delete) {
+        headerValueByName("apikey") { apikey =>
+          auth(apikey) { authPair =>
+            post {
+              addRoute(authPair._1, authPair._2)
+            } ~
+            put {
+              updateRoute(authPair._1, authPair._2)
             } ~
-            path((VendorPattern / NamePattern / FormatPattern / VersionPattern)
-              .repeat(separator = ",")) { list =>
-                val transposed = list.map(_.toList).transpose
-                readRoute(transposed(0), transposed(1), transposed(2),
-                  transposed(3), authPair._1, authPair._2)
+            delete {
+              deleteRoute(authPair._1, authPair._2)
             } ~
-            pathPrefix(VendorPattern.repeat(separator = ",")) { v =>
-              pathPrefix(NamePattern.repeat(separator = ",")) { n =>
-                pathPrefix(FormatPattern.repeat(separator = ",")) { f =>
-                  path(
-                    VersionPattern.repeat(separator = ",")) { vs =>
-                      readRoute(v, n, f, vs, authPair._1, authPair._2)
+            get {
+              path("public") {
+                publicSchemasRoute(authPair._1, authPair._2)
+              } ~
+              path((VendorPattern / NamePattern / FormatPattern / VersionPattern).repeat(1, Int.MaxValue, ",")) { list =>
+                val (vendors, names, formats, versions) = (list.map(_._1), list.map(_._2), list.map(_._3), list.map(_._4))
```

Comment:
Quite "unbeautiful" and inefficient way to extract tuple elements.

Reply:
Updated my [approach](https://github.com/snowplow/iglu/blob/release/r9/2-repositories/scala-repo-server/src/main/scala/com.snowplowanalytics.iglu.server/service/SchemaService.scala#L107-L110). How does it look now?

Reply:
Yep, much better. Though I prefer to avoid `transpose` due its unsafe nature, but that's fine here.